config/api/src/main/java/org/jboss/windup/config/AbstractRuleProvider.java

```diff
@@ -0,0 +1,143 @@
+ */
+package org.jboss.windup.config;
+
+import org.jboss.forge.furnace.util.Annotations;
+import org.jboss.windup.config.metadata.Metadata;
+import org.jboss.windup.config.metadata.MetadataBuilder;
+import org.jboss.windup.config.metadata.RuleMetadata;
+import org.jboss.windup.config.metadata.RuleProviderMetadata;
+import org.jboss.windup.graph.GraphContext;
+import org.ocpsoft.rewrite.config.Rule;
+import org.ocpsoft.rewrite.context.Context;
+import org.ocpsoft.rewrite.context.ContextBase;
+
+/**
+ * {@link AbstractRuleProvider} provides metadata, and a list of {@link Rule} objects that are then evaluated by the
+ * {@link RuleSubet} during Windup execution.
+ *
+ * @author <a href="mailto:lincolnbaxter@gmail.com">Lincoln Baxter, III</a>
+ */
+public abstract class AbstractRuleProvider extends ContextBase implements RuleProvider
+{
+    private int executionIndex;
+    private RuleProviderMetadata metadata;
+
+    public AbstractRuleProvider()
+    {
+        if (!Annotations.isAnnotationPresent(getClass(), Metadata.class))
+        {
+            throw new IllegalStateException(getClass().getName() + " must either specify @" + Metadata.class.getName()
+                        + " or call a constructor and provide " + RuleProviderMetadata.class.getName());
+        }
+        this.metadata = MetadataBuilder.forProvider(getClass());
+    }
+
+    /**
+     * Create a new {@link AbstractRuleProvider} instance using the given {@link RuleProviderMetadata}.
+     */
+    public AbstractRuleProvider(RuleProviderMetadata metadata)
+    {
+        this.metadata = metadata;
+    }
+
+    /**
+     * Create a new {@link AbstractRuleProvider} instance using the given parameters to construct default
+     * {@link RuleProviderMetadata}.
+     */
+    public AbstractRuleProvider(Class<? extends RuleProvider> implementationType, String id)
+    {
+        this.metadata = MetadataBuilder.forProvider(implementationType, id);
+    }
+
+    @Override
+    public RuleProviderMetadata getMetadata()
+    {
+        return metadata;
+    }
+
+    @Override
+    public boolean handles(Object payload)
+    {
+        return payload instanceof GraphContext;
+    }
+
+    /**
+     * Specify additional meta-data to individual {@link Rule} instances originating from the corresponding
+     * {@link RuleProvider} instance.
+     */
+    public static void enhanceRuleMetadata(RuleProvider provider, Rule rule)
+    {
+        if (rule instanceof Context)
+        {
+            Context context = (Context) rule;
+            if (!context.containsKey(RuleMetadata.ORIGIN))
+                context.put(RuleMetadata.ORIGIN, provider.getMetadata().getOrigin());
+            if (!context.containsKey(RuleMetadata.RULE_PROVIDER))
+                context.put(RuleMetadata.RULE_PROVIDER, provider);
+            if (!context.containsKey(RuleMetadata.TAGS))
+                context.put(RuleMetadata.TAGS, provider.getMetadata().getTags());
+        }
```

Comment:
Where's the ultimate storage for the metadata in the end? Is it Context? Does the metadata API load it from there?